### DISCUSSION TITLE
Clear files without DB association

### DISCUSSION THREAD

**Discussion by nuno84** [supabase/supabase#13741] [Questions]
Hi,
I have an avatar url associated with a user.
Whenever I upload a new avatar, it uploads to the supabase storage. All great and easy.
I found that after a while I have a lot of unused images.
What is the best way to keep the bucket trimmed and avoid unused assets? Is there a DB trigger to delete it? Make a sort of "cron" job to check lingering items?
In this case it is an avatar but this question is applicable to all other assets associated with a DB record.
Thank you

**Comment by jpmasters** [comment]
If this is just for avatars, you can solve this by re-using the same avatar path/file name and upserting the new image. That way the old one just gets replaced and it becomes self-managing.

**Comment by GaryAustin1** [comment]
You might be able to write a trigger to use, or adapt my cron code to do what you want. 

https://github.com/GaryAustin1/supa-file-helper

**Comment by HenriqueRamos13** [comment]
In my case, each user have a folder inside the profiles bucket, basically the path is like this: profiles/userId/timestamp.fileType. If i upload a new image i can just delete all images inside the /userId folder. But this is in my case that i know each bucket have only 1 folder per user and each folder have only 1 image. So i delete the old and upload the new with a trigger.

Btw, let me know if this make sense to you.

**Comment by nuno84** [comment]
Hello @GaryAustin1 and @HenriqueRamos13 
@GaryAustin1 thank you for sharing that code, I will take a look. It seems interesting to clean lingering objects.
I think that cleaning on the upload avoids unnecessary processes runing over and over and that genral cleaning cron job can work less frequently.
@HenriqueRamos13 thank you for your suggestion, can you share that trigger?
Thank you

**Reply by HenriqueRamos13** [reply]
is not exactly what i said but is that my code:

```
create or replace function storage.after_insert_objects_on_profiles() 
returns trigger as $$
begin
  update public.users set profile_image = new.name where id = auth.uid();

  return new;
end;
$$ language plpgsql security definer;
create trigger after_insert_objects_on_profiles
  after insert on storage.objects
  for each row
  when (new.bucket_id = 'profiles')
  execute procedure storage.after_insert_objects_on_profiles();

create or replace function storage.before_insert_objects_on_profiles() 
returns trigger as $$
declare
  last_profile_image text;
begin
  select name into last_profile_image from storage.objects where bucket_id = 'profiles' 
    and (storage.foldername(name))[1] = auth.uid()::text order by created_at desc limit 1;

  if last_profile_image is not null and last_profile_image <> new.name then
    delete from storage.objects where bucket_id = 'profiles' and name = last_profile_image;
  end if;

  return new;
end;
$$ language plpgsql security definer;
create trigger before_insert_objects_on_profiles
  before insert on storage.objects
  for each row
  when (new.bucket_id = 'profiles')
  execute procedure storage.before_insert_objects_on_profiles();
```

**Reply by nuno84** [reply]
This seems what I am looking for.
I will take a good look.
Thank you for sharing.

**Reply by GaryAustin1** [reply]
The code above is very bad if I read it correctly.  It appears to delete rows in the storage.objects table.   DO NOT DO THAT.  That orphans the file in s3 storage as you are not calling the storage-api to delete it.  You must either call the storage-api from a client, an edge function or with the pg_net/http extension in a database function.

**Reply by HenriqueRamos13** [reply]
Why exactly is bad? What is let the file orphan and why this is bad? And any example code of using pg_net or a edge function?

**Reply by GaryAustin1** [reply]
The file data is still in s3 storage taking up space (you get billed for it) and there is no way to delete it after your change 
because you deleted the row for it in storage.objects.   The storage-api does not know you deleted that row (no trigger) and so does not know to delete the file from s3.

All the Supabase guides show using the http extension at the end of them for a single file. https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs#storage-management.   The link to my repository above shows a more advanced way to deal with file deletion, but maybe overkill for just deleting an avatar.   It shows how to use http and pg_net.

**Reply by HenriqueRamos13** [reply]
Now I get it! Make sense.

I'm reading your code, but have some things that I didn't get, like:

1:  instead create the new table for config-stuff can I pass it as params to the SQL function and call it on my back-end code? 

2: basically the pg_net is a extension to make https requests right?

And 3: What is the best way to create Cron jobs on supabase?

**Reply by GaryAustin1** [reply]
If you have backend code not clear why you can't just do the storage api calls there.
1) The main function is running in cron so you don't call it.
2) pg_net makes http calls like the http extension.  pg_net is async and written by Supabase.
3) I create a cron task in my code.  You have to run some SQL to do it there is no UI.

**Reply by HenriqueRamos13** [reply]
Hi @GaryAustin1 i was trying to run the code provided by the supabase but it request a projectId on my delete_storage_object function and in local i provided the url: 'http://localhost:54321'. My supabase is running and i tried to delete my profile and i received the error: `{"statusCode":"500","error":"Database Error","message":"Failed to connect to localhost port 54321: Connection refused"}`. Do ou know how to fix it and test on my development local?
